File: Data Structures & Algorithms/reorder-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        // there are two halves of the linked list
        // we want to reverse the second half of the list
        // which we can get to buy using fast and slow pointers
        // where slow.next is the beginning of the second half of the linked list
        // afterward, we traverse both linked lists alternating 
        // and then point the last node to null 

        // fast and slow pointers to get to second half
        ListNode slow = head;
        ListNode fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next; 
        }

        // now we reverse the linked list here
        ListNode second = slow.next;
        slow.next = null;
        ListNode prev = null; // this will become the head of the reversed linked list
        while (second != null) {
            ListNode next = second.next;
            second.next = prev;
            prev = second;
            second = next; 
        }

        // now we can reorder our list by merging both lists
        second = prev; 
        ListNode first = head;
        while (second != null) {
            ListNode temp1 = first.next, temp2 = second.next;
            first.next = second;
            second.next = temp1;
            first = temp1;
            second = temp2;
        }

    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        size = 0
        curr = head
        while curr:
            curr = curr.next
            size += 1
        
        dummy = ListNode(-1)
        dummy.next = head

        prev = dummy
        first = None
        second = head
        for _ in range(size // k):
            prev_group = second
            i = 0
            # reverse this group
            while i < k and second:
                temp = second.next
                second.next = first
                first = second
                second = temp
                i += 1
            
            # go to next group
            prev.next = first
            prev_group.next = second
            prev = prev_group
            first = None
        
        return dummy.next

        # [1, 2, 3, 4, 5, 6] k = 3
        # size = 6, dummy.next -> Node(1)
        # prev = dummy, first = None, second = Node(1)
        # for _ in range(2): runs twice
        # prev_group = Node(1), i = 0
        # while i < k (i = 0) and second:
        # temp = Node(2), second.next = None, first = Node(1), second = Node(2), i = 1
        # temp = Node(3), second.next = Node(1), first = Node(2), second = Node(3), i = 2
        # temp = Node(4), second.next = Node(2), first = Node(3), second = Node(4), i = 3
        # prev = Node(1)
        # prev.next = Node(4)
        # first = none